### PR TITLE
VirtualMachineVolumeExport CRD

### PR DESCRIPTION
**Pull request by hyoung-90**
The VirtualMachineVolumeExport CustomResourceDefinition (CRD) converts the raw format image of the source VirtualMachineVolume to qcow2 image and export it according to the destination.

The VirtualMachineVolumeExport spec's API is split into two parts, the **virtualMachineVolume** and **destination**.
The **virtualMachineVolume** represents the source VirtualMachineVolume to export.
The **destination** represents where the VirtualMachineVolume should be export. Only local destination has been defined.

When destination is local, the VirtualMachineVolumeExport creates a export pvc and save qcow2 image file as **disk.img** in the export pvc. Then the VirtualMachineVolumeExport creates local pod and attaches the export pvc to the local pod to provide to the user.

When destination is object storage(s3), the VirtualMachineVolumeExport will upload qcow2 image to the specified object storage.

### CODE REVIEW

**Comment by woohhan** [comment]
please rebase the commit, and check commit/source has no Korean.

**Comment by hyoung-90** [comment]
> please rebase the commit, and check commit/source has no Korean. @hyoung-90 

Done.

File: pkg/apis/hypercloud/v1alpha1/virtualmachinevolumeexport_types.go [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+// VirtualMachineVolumeExportSpec defines the desired state of VirtualMachineVolumeExport
+type VirtualMachineVolumeExportSpec struct {
+	VirtualMachineVolume VirtualMachineVolumeSource            `json:"virtualMachineVolume"`
+	Destination          VirtualMachineVolumeExportDestination `json:"destination"`
+}
+
+// VirtualMachineVolumeExportDestination defines the destination to export volume
+type VirtualMachineVolumeExportDestination struct {
+	Local *VirtualMachineVolumeExportDestinationLocal `json:"local,omitempty"`
+	S3    *VirtualMachineVolumeExportDestinationS3    `json:"s3,omitempty"`
```
Comment (woohhan) [inline pkg/apis/hypercloud/v1alpha1/virtualmachinevolumeexport_types.go:19]: S3 is not supported yet, so it is unnecessary

File: pkg/apis/hypercloud/v1alpha1/virtualmachinevolumeexport_types.go [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+// VirtualMachineVolumeExportDestinationS3 defines the S3 destination to export volume
+type VirtualMachineVolumeExportDestinationS3 struct {
+	URL    string `json:"url,omitempty"`
+	Secret string `json:"secretRef,omitempty"`
+}
+
+// VirtualMachineVolumeExportStatus defines the observed state of VirtualMachineVolumeExport
+type VirtualMachineVolumeExportStatus struct {
+	// State is the current state of VirtualMachineVolumeExport
+	State VirtualMachineVolumeExportState `json:"state"`
```
Comment (woohhan) [inline pkg/apis/hypercloud/v1alpha1/virtualmachinevolumeexport_types.go:39]: We've decided to use the condition array. Please open the issue to resolve it in another commit

File: pkg/apis/hypercloud/v1alpha1/virtualmachinevolumeexport_types.go [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+}
+
+// VirtualMachineVolumeExportState is the current state of the VirtualMachineVolumeExport
+type VirtualMachineVolumeExportState string
+
+const (
+	// VirtualMachineVolumeExportStateCreating indicates VirtualMachineVolumeExport is creating
+	VirtualMachineVolumeExportStateCreating VirtualMachineVolumeExportState = "Creating"
+	// VirtualMachineVolumeExportStateExporting indicates the exporter pod for VirtualMachineVolumeExport is working
+	VirtualMachineVolumeExportStateExporting VirtualMachineVolumeExportState = "Exporting"
```
Comment (woohhan) [inline pkg/apis/hypercloud/v1alpha1/virtualmachinevolumeexport_types.go:54]: How is `Creating` and `Exporting` different? Do we need both?

File: pkg/apis/hypercloud/v1alpha1/virtualmachinevolumeexport_types.go [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+	// VirtualMachineVolumeExportStateExportCompleted indicates the exporter pod is completed
+	VirtualMachineVolumeExportStateExportCompleted VirtualMachineVolumeExportState = "ExportCompleted"
+	// VirtualMachineVolumeExportStateAvailable indicates VirtualMachineVolumeExport is available
+	VirtualMachineVolumeExportStateAvailable VirtualMachineVolumeExportState = "Available"
```
Comment (woohhan) [inline pkg/apis/hypercloud/v1alpha1/virtualmachinevolumeexport_types.go:58]: Same with prev. What's meaning of `Available` on export?

File: pkg/controller/virtualmachinevolumeexport/exporter_pod.go [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+			return err
+		}
+		if err := r.client.Create(context.Background(), newPod); err != nil && !errors.IsAlreadyExists(err) {
+			return err
+		}
+	}
+	return nil
+}
+
+func getExporterPodName(vmvExportName string) string {
```
Comment (woohhan) [inline pkg/controller/virtualmachinevolumeexport/exporter_pod.go:83]: We need to export this function to use in other code. (and add the comment)

File: pkg/controller/virtualmachinevolumeexport/local_pod.go [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/klog"
+	hc "kubevirt-image-service/pkg/apis/hypercloud/v1alpha1"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+)
+
+func (r *ReconcileVirtualMachineVolumeExport) syncLocalPod() error {
+	completed, found, err := r.isPvcCompleted()
```
Comment (woohhan) [inline pkg/controller/virtualmachinevolumeexport/local_pod.go:17]: Do we need a local Pod? or just not delete the exporter pod after complete. Which one looks better?

File: pkg/controller/virtualmachinevolumeexport/pvc.go [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+	}
+
+	// get virtual machine volume pvc
+	sourcePvc, err := r.getPvc(vmv.GetVolumePvcName(r.vmvExport.Spec.VirtualMachineVolume.Name))
+	if err != nil {
+		return err
+	}
+
+	klog.Infof("Create a new pvc for vmvExport %s", r.vmvExport.Name)
+	if err2 := r.updateStateWithReadyToUse(r.vmvExport, false, hc.VirtualMachineVolumeExportStateCreating); err2 != nil {
```
Comment (woohhan) [inline pkg/controller/virtualmachinevolumeexport/pvc.go:32]: why err2?

File: pkg/controller/virtualmachinevolumeexport/pvc.go [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/klog"
+	hc "kubevirt-image-service/pkg/apis/hypercloud/v1alpha1"
+	vmv "kubevirt-image-service/pkg/controller/virtualmachinevolume"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+)
+
+func (r *ReconcileVirtualMachineVolumeExport) syncPvc() error {
```
Comment (woohhan) [inline pkg/controller/virtualmachinevolumeexport/pvc.go:18]: Please give a more specific name than just `syncPvc`. What does PVC mean?

File: pkg/controller/virtualmachinevolumeexport/exporter_pod.go [hunk 9]
```diff
@@ -0,0 +1,158 @@
...
+	// ExporterImage indicates exporter container image
+	ExporterImage = "quay.io/tmaxanc/kubevirt-image-service-exporter:latest"
+	// ExporterDestinationLocal indicates Destaination to export is local
+	ExporterDestinationLocal = "local"
+	// ExporterDestinationS3 indicates Destaination to export is s3
+	ExporterDestinationS3 = "s3"
+)
+
+func (r *ReconcileVirtualMachineVolumeExport) syncExporterPod() error {
+	completed, found, err := r.isPvcCompleted()
```
Comment (woohhan) [inline pkg/controller/virtualmachinevolumeexport/exporter_pod.go:45]: What does `Complete` mean? We need an explanation(with comment) on this.

File: pkg/controller/virtualmachinevolumeexport/virtualmachinevolumeexport_controller.go [hunk 10]
```diff
@@ -0,0 +1,167 @@
...
+	var destination string
+	if r.vmvExport.Spec.Destination.Local != nil {
+		destination = ExporterDestinationLocal
+	} else if r.vmvExport.Spec.Destination.S3 != nil {
+		destination = ExporterDestinationS3
+	}
+	return destination
+}
+
+// updateStateWithReadyToUse 는 readyToUse와 State를 업데이트 한다. 다른 Status 필드에는 영향을 미치지 않는다. vmi는 캐시를 오염시키지 않기 위해 반드시 DeepCopy 되어 있어야 한다.
```
Comment (woohhan) [inline pkg/controller/virtualmachinevolumeexport/virtualmachinevolumeexport_controller.go:147]: Check the comment

File: pkg/controller/virtualmachinevolumeexport/virtualmachinevolumeexport_controller.go [hunk 11]
```diff
@@ -0,0 +1,167 @@
...
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+/**
+* USER ACTION REQUIRED: This is a scaffold file intended for the user to modify with their own Controller
```
Comment (woohhan) [inline pkg/controller/virtualmachinevolumeexport/virtualmachinevolumeexport_controller.go:21]: Not needed comments

File: pkg/apis/hypercloud/v1alpha1/virtualmachinevolumeexport_types.go (hunk 1 above)
Comment (hyoung-90) [reply pkg/apis/hypercloud/v1alpha1/virtualmachinevolumeexport_types.go]: Done.

File: pkg/controller/virtualmachinevolumeexport/exporter_pod.go (hunk 5 above)
Comment (hyoung-90) [reply pkg/controller/virtualmachinevolumeexport/exporter_pod.go]: Done.

File: pkg/controller/virtualmachinevolumeexport/local_pod.go (hunk 6 above)
Comment (hyoung-90) [reply pkg/controller/virtualmachinevolumeexport/local_pod.go:17]: When the pod is complete, we will not be able to access it.

File: pkg/controller/virtualmachinevolumeexport/pvc.go (hunk 7 above)
Comment (hyoung-90) [reply pkg/controller/virtualmachinevolumeexport/pvc.go:32]: It has lint error.
`shadow: declaration of "err" shadows declaration at line 26`

File: pkg/controller/virtualmachinevolumeexport/virtualmachinevolumeexport_controller.go (hunk 11 above)
Comment (hyoung-90) [reply pkg/controller/virtualmachinevolumeexport/virtualmachinevolumeexport_controller.go]: Done.

File: pkg/controller/virtualmachinevolumeexport/virtualmachinevolumeexport_controller.go (hunk 10 above)
Comment (hyoung-90) [reply pkg/controller/virtualmachinevolumeexport/virtualmachinevolumeexport_controller.go]: Done.

File: pkg/apis/hypercloud/v1alpha1/virtualmachinevolumeexport_types.go (hunk 4 above)
Comment (hyoung-90) [reply pkg/apis/hypercloud/v1alpha1/virtualmachinevolumeexport_types.go]: `Available` means the pvc export is completed, so `ExportCompleted` is better.

File: pkg/apis/hypercloud/v1alpha1/virtualmachinevolumeexport_types.go (hunk 3 above)
Comment (hyoung-90) [reply pkg/apis/hypercloud/v1alpha1/virtualmachinevolumeexport_types.go]: `Creating` is enough.

File: pkg/apis/hypercloud/v1alpha1/virtualmachinevolumeexport_types.go (hunk 4 above)
Comment (woohhan) [reply pkg/apis/hypercloud/v1alpha1/virtualmachinevolumeexport_types.go]: Our CR has the name `export` so `ExportCompleted` is not fit. Just Completed is enough.

File: pkg/controller/virtualmachinevolumeexport/pvc.go (hunk 7 above)
Comment (woohhan) [reply pkg/controller/virtualmachinevolumeexport/pvc.go:32]: That's weird. how about this?
```
if err = r.(); err != nil {

}
```

Comment (hyoung-90) [reply pkg/controller/virtualmachinevolumeexport/pvc.go:32]: I've tried it too, but I still get the lint error.

sloppyReassign: re-assignment to `err` can be replaced with `err := r.updateStateWithReadyToUse(r.vmvExport, false, hc.VirtualMachineVolumeExportStateCreating)` (gocritic)

File: pkg/apis/hypercloud/v1alpha1/virtualmachinevolumeexport_types.go (hunk 4 above)
Comment (hyoung-90) [reply pkg/apis/hypercloud/v1alpha1/virtualmachinevolumeexport_types.go]: Done.

File: pkg/controller/virtualmachinevolumeexport/exporter_pod.go (hunk 9 above)
Comment (hyoung-90) [reply pkg/controller/virtualmachinevolumeexport/exporter_pod.go]: `completed` indicates if pvc export is completed. Comments added.

File: pkg/controller/virtualmachinevolumeexport/pvc.go (hunk 8 above)
Comment (hyoung-90) [reply pkg/controller/virtualmachinevolumeexport/pvc.go]: PVC indicates the volume containing the image file converted from raw format of vmv to be exported to qcow2. How about `syncExportPvc`?

File: pkg/apis/hypercloud/v1alpha1/virtualmachinevolumeexport_types.go (hunk 2 above)
Comment (hyoung-90) [reply pkg/apis/hypercloud/v1alpha1/virtualmachinevolumeexport_types.go:32]: Done.

**Comment by woohhan** [comment]
@hbinkim Please review the commit

**Comment by woohhan** [comment]
test

File: pkg/controller/virtualmachinevolumeexport/virtualmachinevolumeexport_controller.go [hunk 12]
```diff
@@ -0,0 +1,158 @@
...
+func (r *ReconcileVirtualMachineVolumeExport) validateVirtualMachineVolume() error {
+	klog.Infof("Validate vmv for vmvExport %s", r.vmvExport.Name)
+	// check if vmv is exist
+	vmVolume := &hc.VirtualMachineVolume{}
+	err := r.client.Get(context.TODO(), types.NamespacedName{Namespace: r.vmvExport.Namespace, Name: r.vmvExport.Spec.VirtualMachineVolume.Name}, vmVolume)
+	if err != nil {
+		return goerrors.New("fail to get virtual machine volume")
+	}
+	// check if vmv is available
+	if vmVolume.Status.State != hc.VirtualMachineVolumeStateAvailable {
```
Comment (hbinkim) [inline pkg/controller/virtualmachinevolumeexport/virtualmachinevolumeexport_controller.go:154]: I guess this part can be updated later to check vmv condition array

**Comment by hbinkim** [comment]
LGTM

File: pkg/controller/virtualmachinevolumeexport/virtualmachinevolumeexport_controller.go (hunk 12 above)
Comment (woohhan) [reply pkg/controller/virtualmachinevolumeexport/virtualmachinevolumeexport_controller.go:154]: please add to handle this. @young

Comment (hyoung-90) [reply pkg/controller/virtualmachinevolumeexport/virtualmachinevolumeexport_controller.go:154]: If #14 merged, I will update to check vmv condition array. @woohhan

Comment (woohhan) [reply pkg/controller/virtualmachinevolumeexport/virtualmachinevolumeexport_controller.go:154]: we need an issue for tracking

File: pkg/apis/hypercloud/v1alpha1/virtualmachinevolumeexport_types.go [hunk 13]
```diff
@@ -0,0 +1,80 @@
+package v1alpha1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// EDIT THIS FILE!  THIS IS SCAFFOLDING FOR YOU TO OWN!
```
Comment (woohhan) [inline pkg/apis/hypercloud/v1alpha1/virtualmachinevolumeexport_types.go:7]: not needed comments

File: pkg/controller/virtualmachinevolumeexport/virtualmachinevolumeexport_controller.go (hunk 12 above)
Comment (hyoung-90) [reply pkg/controller/virtualmachinevolumeexport/virtualmachinevolumeexport_controller.go:154]: #19 is related.

File: pkg/apis/hypercloud/v1alpha1/virtualmachinevolumeexport_types.go (hunk 13 above)
Comment (hyoung-90) [reply pkg/apis/hypercloud/v1alpha1/virtualmachinevolumeexport_types.go]: Done.

**Pull request merged**